File: tests/rosetta/transpiler/FS/9-billion-names-of-god-the-integer.fs
// Generated 2025-07-24 15:11 +0700

exception Return

let rec bigTrim (a: int array) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable a = a
    try
        let mutable n: int = Seq.length a
        while (n > 1) && ((a.[n - 1]) = 0) do
            a <- Array.sub a 0 ((n - 1) - 0)
            n <- n - 1
        __ret <- a
        raise Return
        __ret
    with
        | Return -> __ret
and bigFromInt (x: int) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable x = x
    try
        if x = 0 then
            __ret <- [|0|]
            raise Return
        let mutable digits: int array = [||]
        let mutable n: int = x
        while n > 0 do
            digits <- Array.append digits [|n % 10|]
            n <- n / 10
        __ret <- digits
        raise Return
        __ret
    with
        | Return -> __ret
and bigAdd (a: int array) (b: int array) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable a = a
    let mutable b = b
    try
        let mutable res: int array = [||]
        let mutable carry: int = 0
        let mutable i: int = 0
        while ((i < (Seq.length a)) || (i < (Seq.length b))) || (carry > 0) do
            let mutable av: int = 0
            if i < (Seq.length a) then
                av <- a.[i]
            let mutable bv: int = 0
            if i < (Seq.length b) then
                bv <- b.[i]
            let mutable s = (av + bv) + carry
            res <- Array.append res [|s % 10|]
            carry <- s / 10
            i <- i + 1
        __ret <- bigTrim res
        raise Return
        __ret
    with
        | Return -> __ret
and bigSub (a: int array) (b: int array) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable a = a
    let mutable b = b
    try
        let mutable res: int array = [||]
        let mutable borrow: int = 0
        let mutable i: int = 0
        while i < (Seq.length a) do
            let mutable av = a.[i]
            let mutable bv: int = 0
            if i < (Seq.length b) then
                bv <- b.[i]
            let mutable diff = (av - bv) - borrow
            if diff < 0 then
                diff <- diff + 10
                borrow <- 1
            else
                borrow <- 0
            res <- Array.append res [|diff|]
            i <- i + 1
        __ret <- bigTrim res
        raise Return
        __ret
    with
        | Return -> __ret
and bigToString (a: int array) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable a = a
    try
        let mutable s: string = ""
        let mutable i: int = (Seq.length a) - 1
        while i >= 0 do
            s <- s + (string (a.[i]))
            i <- i - 1
        __ret <- s
        raise Return
        __ret
    with
        | Return -> __ret
and minInt (a: int) (b: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable a = a
    let mutable b = b
    try
        if a < b then
            __ret <- a
            raise Return
        else
            __ret <- b
            raise Return
        __ret
    with
        | Return -> __ret
and cumu (n: int) =
    let mutable __ret : int array array = Unchecked.defaultof<int array array>
    let mutable n = n
    try
        let mutable cache: int array array array = [|[|bigFromInt 1|]|]
        let mutable y: int = 1
        while y <= n do
            let mutable row: int array array = [|bigFromInt 0|]
            let mutable x: int = 1
            while x <= y do
                let ``val`` = cache.[y - x].[minInt x (y - x)]
                row <- Array.append row [|bigAdd (row.[(Array.length row) - 1]) ``val``|]
                x <- x + 1
            cache <- Array.append cache [|row|]
            y <- y + 1
        __ret <- cache.[n]
        raise Return
        __ret
    with
        | Return -> __ret
and row (n: int) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable n = n
    try
        let e = cumu n
        let mutable out: string array = [||]
        let mutable i: int = 0
        while i < n do
            let diff = bigSub (e.[i + 1]) (e.[i])
            out <- Array.append out [|bigToString diff|]
            i <- i + 1
        __ret <- out
        raise Return
        __ret
    with
        | Return -> __ret
printfn "%s" "rows:"
let mutable x: int = 1
while x < 11 do
    let r = row x
    let mutable line: string = ""
    let mutable i: int = 0
    while i < (Seq.length r) do
        line <- ((line + " ") + (r.[i])) + " "
        i <- i + 1
    printfn "%s" line
    x <- x + 1
printfn "%s" ""
printfn "%s" "sums:"
for num in [|23; 123; 1234|] do
    let r = cumu num
    printfn "%s" (((string num) + " ") + (bigToString (r.[(Seq.length r) - 1])))

File: tests/rosetta/transpiler/Clojure/9-billion-names-of-god-the-integer.clj
(ns main (:refer-clojure :exclude [bigTrim bigFromInt bigAdd bigSub bigToString minInt cumu row]))

(require 'clojure.set)

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare bigTrim bigFromInt bigAdd bigSub bigToString minInt cumu row)

(defn bigTrim [a]
  (try (do (def n (count a)) (while (and (> n 1) (= (nth a (- n 1)) 0)) (do (def a (subvec a 0 (- n 1))) (def n (- n 1)))) (throw (ex-info "return" {:v a}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn bigFromInt [x]
  (try (do (when (= x 0) (throw (ex-info "return" {:v [0]}))) (def digits []) (def n x) (while (> n 0) (do (def digits (conj digits (mod n 10))) (def n (/ n 10)))) (throw (ex-info "return" {:v digits}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn bigAdd [a b]
  (try (do (def res []) (def carry 0) (def i 0) (while (or (or (< i (count a)) (< i (count b))) (> carry 0)) (do (def av 0) (when (< i (count a)) (def av (nth a i))) (def bv 0) (when (< i (count b)) (def bv (nth b i))) (def s (+ (+ av bv) carry)) (def res (conj res (mod s 10))) (def carry (/ s 10)) (def i (+ i 1)))) (throw (ex-info "return" {:v (bigTrim res)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn bigSub [a b]
  (try (do (def res []) (def borrow 0) (def i 0) (while (< i (count a)) (do (def av (nth a i)) (def bv 0) (when (< i (count b)) (def bv (nth b i))) (def diff (- (- av bv) borrow)) (if (< diff 0) (do (def diff (+ diff 10)) (def borrow 1)) (def borrow 0)) (def res (conj res diff)) (def i (+ i 1)))) (throw (ex-info "return" {:v (bigTrim res)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn bigToString [a]
  (try (do (def s "") (def i (- (count a) 1)) (while (>= i 0) (do (def s (str s (str (nth a i)))) (def i (- i 1)))) (throw (ex-info "return" {:v s}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn minInt [a b]
  (if (< a b) (throw (ex-info "return" {:v a})) (throw (ex-info "return" {:v b}))))

(defn cumu [n]
  (try (do (def cache [[(bigFromInt 1)]]) (def y 1) (while (<= y n) (do (def row_v [(bigFromInt 0)]) (def x 1) (while (<= x y) (do (def val (nth (nth cache (- y x)) (minInt x (- y x)))) (def row_v (conj row_v (bigAdd (nth row_v (- (count row_v) 1)) val))) (def x (+ x 1)))) (def cache (conj cache row_v)) (def y (+ y 1)))) (throw (ex-info "return" {:v (nth cache n)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn row [n]
  (try (do (def e (cumu n)) (def out []) (def i 0) (while (< i n) (do (def diff (bigSub (nth e (+ i 1)) (nth e i))) (def out (conj out (bigToString diff))) (def i (+ i 1)))) (throw (ex-info "return" {:v out}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(def x 1)

(defn -main []
  (println "rows:")
  (while (< x 11) (do (def r (row x)) (def line "") (def i 0) (while (< i (count r)) (do (def line (str (str (str line " ") (nth r i)) " ")) (def i (+ i 1)))) (println line) (def x (+ x 1))))
  (println "")
  (println "sums:")
  (doseq [num [23 123 1234]] (do (def r (cumu num)) (println (str (str (str num) " ") (bigToString (nth r (- (count r) 1))))))))

(-main)
